pkg/console/service/service.go

```diff
@@ -0,0 +1,353 @@
+				Resources:     []string{"virtualmachineinstances/vnc"},
+				ResourceNames: []string{vmMeta.GetName()},
+				Verbs:         []string{"get"},
+			}}
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create role: %w", err)
+	}
+
+	_, err = createOrUpdate[*rbac.RoleBinding](
+		ctx,
+		name,
+		namespace,
+		s.kubevirtClient.RbacV1().RoleBindings(namespace),
+		func(foundObj *rbac.RoleBinding) {
+			foundObj.Labels = commonLabels
+			foundObj.OwnerReferences = []metav1.OwnerReference{vmOwnerRef}
+			foundObj.Subjects = []rbac.Subject{{
+				Kind: "ServiceAccount",
+				Name: serviceAccount.Name,
+			}}
+			foundObj.RoleRef = rbac.RoleRef{
+				APIGroup: rbac.GroupName,
+				Kind:     "Role",
+				Name:     role.Name,
+			}
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create role binding: %w", err)
+	}
+	return nil
+}
+
+func (s *service) requestToken(ctx context.Context, serviceAccountNamespace string, serviceAccountName string, duration time.Duration) (string, error) {
+	durationSeconds := int64(duration.Seconds())
+	tokenRequest := &authnv1.TokenRequest{
+		Spec: authnv1.TokenRequestSpec{
+			Audiences:         nil,
+			ExpirationSeconds: &durationSeconds,
+			BoundObjectRef:    nil,
+		},
+	}
+
+	tokenRequest, err := s.kubevirtClient.CoreV1().ServiceAccounts(serviceAccountNamespace).CreateToken(
+		ctx,
+		serviceAccountName,
+		tokenRequest,
+		metav1.CreateOptions{},
+	)
+	if err != nil {
+		return "", err
+	}
+	return tokenRequest.Status.Token, nil
+}
+
+func getAuthToken(request *restful.Request) string {
+	authHeader := request.HeaderParameter("Authorization")
+
+	const bearerPrefix = "Bearer "
+	if strings.HasPrefix(authHeader, bearerPrefix) {
+		return authHeader[len(bearerPrefix):]
+	}
+	return ""
+}
+
+type tokenRequestParams struct {
+	namespace string
+	name      string
+	duration  time.Duration
+}
+
+func readTokenRequestParameters(request *restful.Request) (*tokenRequestParams, error) {
+	namespace := request.PathParameter("namespace")
+	name := request.PathParameter("name")
+	if namespace == "" || name == "" {
+		return nil, fmt.Errorf("namespace and name parameters are required")
+	}
+
+	duration := 10 * time.Minute
+	durationParam := request.QueryParameter("duration")
+	if durationParam != "" {
+		var err error
+		duration, err = time.ParseDuration(durationParam)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse duration: %w", err)
+		}
+	}
+
+	return &tokenRequestParams{
+		namespace: namespace,
+		name:      name,
+		duration:  duration,
+	}, nil
+}
+
+type clientInterface[PT any] interface {
+	Create(ctx context.Context, obj PT, opts metav1.CreateOptions) (PT, error)
+	Update(ctx context.Context, obj PT, opts metav1.UpdateOptions) (PT, error)
+	Delete(ctx context.Context, name string, opts metav1.DeleteOptions) error
+	Get(ctx context.Context, name string, opts metav1.GetOptions) (PT, error)
+}
+
+func createOrUpdate[PT interface {
```

Comment:
Could this and L332 be part of a library? I see those kind of functions duplicated a lot across our projects.

Reply:
The `retryOnConflict()` function is nearly a copy of this one from `k8s.io/client-go`: https://github.com/kubernetes/client-go/blob/64a35f6a46ec8a791d437495fd91c87bcd01a5b5/util/retry/util.go#L48-L66

Where are they duplicated?